spec/unit/typesafe/config/config_value_spec.rb

```diff
@@ -0,0 +1,506 @@
+    let(:second_object) { a }
+    include_examples "object_equality"
+  end
+
+  context "a equals same_as_a" do
+    let(:first_object) { a }
+    let(:second_object) { same_as_a }
+    include_examples "object_equality"
+  end
+
+  context "a doesn't equal b" do
+    let(:first_object) { a }
+    let(:second_object) { b }
+    include_examples "object_inequality"
+  end
+end
+
+describe "ConfigObject" do
+  specify "should unwrap correctly" do
+    m = SimpleConfigObject.new(TestUtils.fake_origin, TestUtils.config_map({a: 1, b: 2, c: 3}))
+
+    expect({a: 1, b: 2, c: 3}).to eq(m.unwrapped)
+  end
+
+  specify "should implement read only map" do
+    m = SimpleConfigObject.new(TestUtils.fake_origin, TestUtils.config_map({a: 1, b: 2, c: 3}))
+
+    expect(TestUtils.int_value(1)).to eq(m[:a])
+    expect(TestUtils.int_value(2)).to eq(m[:b])
+    expect(TestUtils.int_value(3)).to eq(m[:c])
+    expect(m[:d]).to be_nil
+    # [] can take a non-string
+    expect(m[[]]).to be_nil
+
+    expect(m.has_key? :a).to be_truthy
+    expect(m.has_key? :z).to be_falsey
+    # has_key? can take a non-string
+    expect(m.has_key? []).to be_falsey
+
+    expect(m.has_value? TestUtils.int_value(1)).to be_truthy
+    expect(m.has_value? TestUtils.int_value(10)).to be_falsey
+    # has_value? can take a non-string
+    expect(m.has_value? []).to be_falsey
+
+    expect(m.empty?).to be_falsey
+
+    expect(m.size).to eq(3)
+
+    values = [TestUtils.int_value(1), TestUtils.int_value(2), TestUtils.int_value(3)]
+    expect(values).to eq(m.values)
+
+    keys = [:a, :b, :c]
+    expect(keys).to eq(m.keys)
+
+    expect { m["hello"] = TestUtils.int_value(41) }.to raise_error(NoMethodError)
+    expect { m.delete(:a) }.to raise_error(NoMethodError)
+  end
+end
+
+describe "ConfigList" do
+  specify "should implement read only list" do
+    values = ["a", "b", "c"].map { |i| TestUtils.string_value(i) }
+    l = SimpleConfigList.new(TestUtils.fake_origin, values)
+
+    expect(values[0]).to eq(l[0])
+    expect(values[1]).to eq(l[1])
+    expect(values[2]).to eq(l[2])
+
+    expect(l.include? TestUtils.string_value("a")).to be_truthy
+    expect(l.include_all?([TestUtils.string_value("a")])).to be_truthy
+    expect(l.include_all?([TestUtils.string_value("b")])).to be_truthy
+    expect(l.include_all?(values)).to be_truthy
+
+    expect(l.index(values[1])).to eq(1)
+
+    expect(l.empty?).to be_falsey
+
+    expect(l.map { |v| v }).to eq(values.map { |v| v })
+
+    expect(l.rindex(values[1])).to eq(1)
+
+    expect(l.size).to eq(3)
+
+    expect { l.push(TestUtils.int_value(3)) }.to raise_error(NoMethodError)
+    expect { l << TestUtils.int_value(3) }.to raise_error(NoMethodError)
+    expect { l.clear }.to raise_error(NoMethodError)
+    expect { l.delete(TestUtils.int_value(2)) }.to raise_error(NoMethodError)
+    expect { l.delete(1) }.to raise_error(NoMethodError)
+    expect { l[0] = TestUtils.int_value(42) }.to raise_error(NoMethodError)
```

Comment:
After thinking about this some more, I don't know if I actually care that much if we explicitly throw `UnsupportedOperationExceptions` (or something similar) in the ruby code.  The `NoMethodErrors` are probably fine.  It's not something that should ever happen anyway.